sw/host/opentitanlib/src/transport/ti50emulator/gpio.rs

```diff
@@ -241,148 +204,86 @@ impl Ti50GpioPin {
     fn update_dut_state(&self) -> Result<()> {
         if let Some(ref mut pin_fd) = *self.socket.borrow_mut() {
             let mut buf: [u8; GPIO_BUF_SIZE] = [0; GPIO_BUF_SIZE];
-            pin_fd.write_all(&[GpioStreamCommand::GetState as u8])?;
+            pin_fd.write_all(&[b'?'])?;
```

Comment:
Right, let me define a constant (the `GpioStreamCommand` type does not exist anymore, since I had hoped that `Logic` would replace it, which it does except for this one character).

Longer term, I hope that this process will never need to send '?', as it should keep track of the unsolicited characters received from the sub-process, and remember the last one.  This requires some enhancements to make sure that GPIO pipe connections are established early on startup, before the sub-process could make any changes, that is before any ti50 code runs in the sub-process.

We are starting to rely on the `GpioMonitoring` trait to a larger extent in the Ti50 tests, for watching GPIO signals like a logic analyzer.  In order for this to work in host emulation, all the GPIO pipes need to be continuously monitored with `mio`-style polling, in order to timestamp every unsolicited state change.

Reply:
Yes GpioMonitoring is good idea. I was thinking about it when writing this code. that was the reason why he used the word stream in the name of enum. 

> This requires some enhancements to make sure that GPIO pipe connections are established early on startup, before the sub-process could make any changes, that is before any ti50 code runs in the sub-process.

For this reason, "READY" signal checking and a "control socket" have been implemented. I don't know what the current implementation status is, I think you removed it. I believe that in the linked PR you are only checking the text from stdout.
https://github.com/lowRISC/opentitan/pull/17797/files

Reply:
Yes, I want to do something like that, but I will save that for a separate PR that introduces an implementation of `GpioMonitoring`.

One consideration is that the host emulation binary is actually already being used for ChromeOS testing without this opentitansession parent process, and this limits how much we can change about the GPIO socket setup sequence.  I like the approach of opentitansession creating all the GPIO sockets, and the HE process connecting to them, though.  Maybe command line arguments can be used to allow the HE process to distinguish between the two modes.